primitives/src/cmpext.rs

```diff
@@ -0,0 +1,144 @@
+
+/// Type which adds two additional values for any type:
+/// - `Min` which is smaller that any value of this type
+/// - `Max` which is greater that any value of this type
+///
+/// Used to enable query over prefix of the given key in the b-tree e.g. by account id in the asset id.
+///
+/// Suppose compound key of three parts: `K = (A, B, C)`.
+/// So that in sorting order keys will be sorted firstly by `A` then `B` and `C`.
+/// This keys are stored in `BTreeMap` and it's required to extract all keys which have `A == a`.
+/// To do this it's possible to use `range` provided by `BTreeMap`,
+/// but it would't be enough to simply use `(a..=a)` bound for `K` because ranges bounds are found by binary search
+/// and this way any key which has `A == a` can be treated as bound.
+/// So `MinMaxExt` is used to express precise bound for such query: `(a, MIN, MIN)..(a, MAX, MAX)`.
+#[derive(Debug, Clone, Copy)]
+pub enum MinMaxExt<T> {
+    /// Value that is greater than any value
+    Min,
+    /// Value that is smaller than any value
+    Max,
+    /// Regular value
+    Value(T),
+}
+
+impl<T: PartialEq> PartialEq for MinMaxExt<T> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Value(lhs), Self::Value(rhs)) => lhs.eq(rhs),
+            (Self::Min, Self::Min) | (Self::Max, Self::Max) => true,
+            _ => false,
+        }
+    }
+}
+
+impl<T: Eq> Eq for MinMaxExt<T> {}
+
+impl<T: PartialOrd> PartialOrd for MinMaxExt<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match (self, other) {
+            (Self::Value(lhs), Self::Value(rhs)) => lhs.partial_cmp(rhs),
+            (lhs, rhs) if lhs == rhs => Some(Ordering::Equal),
+            (Self::Min, _) | (_, Self::Max) => Some(Ordering::Less),
+            (Self::Max, _) | (_, Self::Min) => Some(Ordering::Greater),
+        }
+    }
+}
+
+impl<T: Ord> Ord for MinMaxExt<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self, other) {
+            (Self::Value(lhs), Self::Value(rhs)) => lhs.cmp(rhs),
+            (lhs, rhs) if lhs == rhs => Ordering::Equal,
+            (Self::Min, _) | (_, Self::Max) => Ordering::Less,
+            (Self::Max, _) | (_, Self::Min) => Ordering::Greater,
+        }
+    }
+}
+
+impl<T> From<T> for MinMaxExt<T> {
+    fn from(value: T) -> Self {
+        MinMaxExt::Value(value)
+    }
+}
+
+/// Helper macro to enable cast of key to dyn object and derive required traits for it.
+/// Used to bypass limitation of [`Borrow`] which wouldn't allow to create object and return reference to it.
+#[macro_export]
+macro_rules! impl_as_dyn_key {
+    (target: $ty:ident, key: $key:ty, trait: $trait:ident) => {
+        /// Trait to key from type
+        pub trait $trait {
+            /// Extract key
+            fn as_key(&self) -> $key;
+        }
+
+        impl $trait for $key {
+            fn as_key(&self) -> $key {
+                *self
+            }
+        }
```

Comment:
I still don't understand how does this work... How can you create a movable value from a reference... It should be possible only if `$key` implements `Copy`

Reply:
Oh, I see, your `RoleIdByAccount` actualy derives `Copy`

Reply:
Was it intenional that this macro works properly only with `$key` implementing `Copy`?

Reply:
> Was it intenional that this macro works properly only with $key implementing Copy?

Yeah, key expected to be light weighted value consisting of references or values which can be copied.